pkg/controllermanager/controller/credentialsbinding/referencecleaner/reconciler.go

```diff
@@ -0,0 +1,152 @@
+		objectsToClean = map[objectId]client.Object{}
+	)
+
+	for _, resource := range []struct {
+		schemaGroupVersion schema.GroupVersion
+		kind               string
+		listKind           string
+	}{
+		{corev1.SchemeGroupVersion, "Secret", "SecretList"},
+		{securityv1alpha1.SchemeGroupVersion, "WorkloadIdentity", "WorkloadIdentityList"},
+	} {
+		objList := &metav1.PartialObjectMetadataList{}
+		objList.SetGroupVersionKind(resource.schemaGroupVersion.WithKind(resource.listKind))
+		if err := r.Reader.List(ctx, objList, labels); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		for _, obj := range objList.Items {
+			obj.SetGroupVersionKind(resource.schemaGroupVersion.WithKind(resource.kind))
+			objectsToClean[objectId{resource.kind, obj.Namespace, obj.Name}] = &obj
+		}
+	}
+
+	credentialsBindingList := &securityv1alpha1.CredentialsBindingList{}
+	if err := r.Reader.List(ctx, credentialsBindingList); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	for _, cb := range credentialsBindingList.Items {
+		delete(objectsToClean, objectId{cb.CredentialsRef.Kind, cb.CredentialsRef.Namespace, cb.CredentialsRef.Name})
+	}
+
+	var (
+		results = make(chan error, 1)
+		wg      wait.Group
+		errList error
+	)
+
+	for id, obj := range objectsToClean {
+		wg.StartWithContext(ctx, func(ctx context.Context) {
+			// Remove shoot provider label and 'referred by a credentials binding' label
+			hasProviderLabel, providerLabel := getProviderLabel(obj.GetLabels())
+			_, hasCredentialsBindingRefLabel := obj.GetLabels()[v1beta1constants.LabelCredentialsBindingReference]
+			_, hasSecretBindingRefLabel := obj.GetLabels()[v1beta1constants.LabelSecretBindingReference]
+			gvk := obj.GetObjectKind().GroupVersionKind()
+			if hasProviderLabel || hasCredentialsBindingRefLabel {
+				patch := client.MergeFrom(obj.DeepCopyObject().(client.Object))
+
+				labels := obj.GetLabels()
+				delete(labels, v1beta1constants.LabelCredentialsBindingReference)
+
+				// The credential can be still referenced by a secretbinding so
+				// only remove the provider label if there is no secretbinding reference label
+				if !hasSecretBindingRefLabel {
+					delete(labels, providerLabel)
+				}
+
+				obj.SetLabels(labels)
+				log.Info("Removing referred label", id.kind, client.ObjectKeyFromObject(obj)) //nolint:logcheck
+				if err := r.Writer.Patch(ctx, obj, patch); err != nil {
+					results <- fmt.Errorf("failed to remove referred label from %s: %w", id.kind, err)
+				}
+			}
+
+			// Remove finalizer from referenced credential
+			if controllerutil.ContainsFinalizer(obj, gardencorev1beta1.ExternalGardenerName) && !hasSecretBindingRefLabel {
+				log.Info("Removing finalizer", id.kind, client.ObjectKeyFromObject(obj)) //nolint:logcheck
+				obj.GetObjectKind().SetGroupVersionKind(gvk)
+				if err := controllerutils.RemoveFinalizers(ctx, r.Writer, obj, gardencorev1beta1.ExternalGardenerName); err != nil {
```

Comment:
Is it possible to send single patch request that removes the labels and finalizer at once? Obviously `RemoveFinalizers` cannot be used in this case, but controller runtime provides function that removes finalizer from object.